### PR TITLE
Add a non-recursive version of label_node.

### PR DESCRIPTION
**Pull request by nh13**
It currently can run out of stack space with too many recursions.

### CODE REVIEW

**Comment by nh13** [comment]
@pjedge it's likely a bug or pathological condition that is causing me to run out of stack space, so I'll add that if I find it.

**Comment by nh13** [comment]
@pjedge I have a small (confidential) test case I can send.  I ran `extractHAIRS` with `--indels 1`, and when I changed it to `--indels 0`, `HAPCUT2` worked.

**Comment by nh13** [comment]
@pjedge the second commit adds support for checking if we have visited a node or not in the DFS when labeling nodes.  If we have very large connected components we may never stop searching!

File: hapcut2-src/fragmatrix.c [hunk 1]
```diff
@@ -1,18 +1,69 @@
...
+                nodes = (int*)realloc(nodes, sizeof(int)*m_nodes);
+            }
+            nodes[n_nodes] = cur_node;
+            n_nodes++;
+            kh_put(32, label_node_hash, cur_node, &ret);
+        }
+    }
+
+    // destroy
+    free(nodes);
```
Comment (nh13) [inline hapcut2-src/fragmatrix.c]: TODO: add `kh_destroy(32, label_node_hash)` so we don't get a memory leak.

File: hapcut2-src/fragmatrix.c [hunk 2]
```diff
@@ -1,18 +1,79 @@
...
+    nodes[0] = init_node;
+    n_nodes  = 1;
+	kh_put(32, label_node_hash, init_node, &ret);
+
+    while (0 < n_nodes) {
+        // get the next node
+        int node = nodes[n_nodes-1];
+        n_nodes--;
+
+        if (snpfrag[node].component == -1) continue;
```
Comment (nh13) [inline hapcut2-src/fragmatrix.c]: I think this should be `!=`

**Review by pjedge — APPROVED** [review]

**Pull request merged**